File: NamasteDSA/Arrays/SingleNumber.java
public class SingleNumber {
    public static int singleNumber(int[] array) {
        int xor = 0;

        for (int index = 0; index < array.length; index++) {
            xor = xor ^ array[index];
        }

        return xor;
    }

    public static void main(String[] args) {
        int[] array1 = { 2, 2, 1 };
        int[] array2 = { 4, 1, 2, 1, 2 };

        System.out.println(singleNumber(array1));
        System.out.println(singleNumber(array2));
    }
}

File: NamasteDSA/Arrays/singleNumber.js
function singleNumber(array) {
  let hash = {};

  for (let i = 0; i < array.length; i++) {
    if (hash[array[i]]) {
      hash[array[i]]++;
    } else {
      hash[array[i]] = 1;
    }
  }

  for (let i = 0; i < array.length; i++) {
    if (hash[array[i]] === 1) {
      return array[i];
    }
  }
}

function singleNumberXOR(array) {
  let xor = 0;

  for (let i = 0; i < array.length; i++) {
    xor = xor ^ array[i];
  }

  return xor;
}

console.log(singleNumberXOR([2, 2, 1]));
console.log(singleNumberXOR([4, 1, 2, 1, 2]));
